[MRG+2] Added warm_start to bagging

betatim opened the pull request:
BaggingClassifier and BaggingRegressor now support warm_starts. Added
basic tests and documentation of the new functionality. Heavily
inspired by work on warm_start for Random forests.

There is a brief discussion of this in #3419 but it seemed dead?

Questions:
- How best to document that the oob score is only calculated on the last batch of training samples?
- What other invariants exist between cold and warm started classifiers that are worth testing?

betatim committed:
Added warm_start to bagging

BaggingClassifier and BaggingRegressor now support warm_starts. Added
basic tests and documentation of the new functionality. Heavily
inspired by work on warm_start for Random forests.

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -454,14 +483,14 @@ def _validate_estimator(self):
-        for estimator, samples, features in zip(self.estimators_,
+        for estimator, samples, features in zip(self.estimators_[-n_more_estimators:],
                                                 self.estimators_samples_,
                                                 self.estimators_features_):
```
looks like there is a bug here.

betatim replied:
estimators_samples_ and estimators_features_ are for the last used batch of samples in `fit()` and estimators_ contains all estimators trained so far. So you have to select only the last `n_more_estimators`

arjoly replied:
Sorry, but I don't get it given the initalisation of `predictions = np.zeros((n_samples, n_classes_))`.

glouppe replied:
`self.estimators_[-n_more_estimators:], self.estimators_samples_` and `self.estimators_features_` are all of the same size. 
- `self.estimators_[-n_more_estimators:]` contains the estimators trained during the last batch.
- `self.estimators_samples_` and `self.estimators_features_` contains the samples (resp. features) indices used during the last batch. 
Then the oob score is computed from the last batch only, as currently done in the forest as well. 

Why do think this is a bug @arjoly ?

I agree however that it may be nice to accumulate oob scores over the batches, but this would require a different logic from what we have right now (here and in forests).

arjoly replied:
> I agree however that it may be nice to accumulate oob scores over the batches, but this would require a different logic from what we have right now (here and in forests).

This is the behaviour that i expect.

glouppe replied:
Okay, then do you agree to leave like this for now, for consistency with how it is done in forests + create an issue to have a better API?

arjoly replied:
Ok, then let's leave this for now.  Are you ok with the erase of estimators_samples_ and estimators_features_ attribute.

glouppe replied:
Meh, I just checked `forest.py` and it seems `_set_oob_score` in forest is wrong upon warm restart. It reuses `X, y` from the last batch, making the indices invalid.

arjoly replied:
I would be in favor of raising an error for now when you want the oob scores with `warm_start=True` and this is not the first batch.

> Meh, I just checked forest.py and it seems _set_oob_score in forest is wrong upon warm restart. It reuses X, y from the last batch, making the indices invalid.

Arf :-/

betatim replied:
What should the `oob_score_` and `oob_decision_function_` attributes be set to after fitting the second batch in order to signal "this attribute isn't available".

Normally the attributes simply don't exist if `oob_score=False`. Should we set it to `None`, leave it set to the value from the first batch, apply some python black magic to remove the attribute again? I dislike the black magic idea, leaving it set to the value form the first batch will potentially confuse people, so `None`?

arjoly replied:
If `warm_start=True` and `oob_score=True`, I would raise a `ValueError`. If was oob_score was previously set, I would erase them (`del`).

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -838,13 +874,13 @@ def _validate_estimator(self):
         n_predictions = np.zeros((n_samples,))
 
-        for estimator, samples, features in zip(self.estimators_,
+        for estimator, samples, features in zip(self.estimators_[-n_more_estimators:],
```
looks like there is a bug here.

betatim replied:
`estimators_samples_` and `estimators_features_` are for the last used batch of samples in `fit()` and `estimators_` contains all estimators trained so far. So you have to select only the last `n_more_estimators`

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -276,27 +278,47 @@ def fit(self, X, y, sample_weight=None):
-        self.estimators_ = list(itertools.chain.from_iterable(
+        self.estimators_ += list(itertools.chain.from_iterable(
             t[0] for t in all_results))
         self.estimators_samples_ = list(itertools.chain.from_iterable(
```
Why not updating  `self.estimators_samples_`here instead of earasing the whole  attribute.

arjoly replied:
It's the same for `estimators_features_`

betatim replied:
With `oob_score_` calculation out of the picture, should we keep `estimators_samples_` and `estimators_features_` for previously fitted estimators? They are useful if you repeatedly call `fit()` with the same `X` and `y`, adding more estimators. They are useless if you are using batches of data. Which is more common?

arjoly commented:
The warm_start with oob_score seems to be untested.

betatim committed:
Fix random_state to make test reproducible

betatim committed:
Test for unchanged n_estimator and indentation fix

Added a test to check nothing happens when calling
fit() without changing n_estimators. Fixed indentation
in fit() for this case.

betatim committed:
Test oob_score for warm_start'ed bagging classifiers

Test that oob_score is a good estimate of generalization
error in case of warm_start.

Test that oob_score is computed only on most recent
training samples with newly added classifiers

betatim commented:
Good point on not resetting `self.estimators_samples_` and `estimators_features_`. Now they aren't reset at the start anymore which was redundant.

Added a test of the oob score when using `warm_start`.

I also found and fixed a mistake in my indentation for the case where `n_estimators` does not change between two calls to `fit()`. Added a test that checks that the classifier's predictions do not change if you call `fit()` without changing `n_estimators`.

amueller commented on sklearn/ensemble/bagging.py:
```diff
@@ -276,40 +278,59 @@ def fit(self, X, y, sample_weight=None):
-                seeds[starts[i]:starts[i + 1]],
-                verbose=self.verbose)
-            for i in range(n_jobs))
+        if n_more_estimators < 0:
```
I'd just test for ``n_more_estimators <= 0`` and raise in either case. Nothing happens if ``n_more_estimators == 0``, right?

betatim replied:
For `n_more_estimators < 0` an exception is raised and for `n_more_estimators == 0` there is just a warning. To me that signals that using <0 is incorrect, and ==0 is not wrong but silly to do. I copied this behaviour from [forest.py](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/forest.py#L242). I have no strong feelings on whether both should raise or not, I copied forest.py's behaviour in the name of consistency.

amueller replied:
Ok, let's leave it like the, then.

amueller commented on sklearn/ensemble/bagging.py:
```diff
@@ -276,40 +278,59 @@ def fit(self, X, y, sample_weight=None):
+            warn("Warm-start fitting without increasing n_estimators does not "
+                 "fit new trees.")
+
+        else:
```
I wouldn't do this ``else`` as it makes all the code be indented more.

betatim replied:
Is it Ok to `return self` after the warning? That is the only way I can think of to get rid of the indentation.

amueller replied:
yeah thats what I meant. Sorry for being unclear. But it is not strictly necessary ;)

glouppe commented:
It looks good to me. Tests are thorough.

+1 for merge.

betatim committed:
Removed overly deep indentation

betatim commented:
I disliked the extra indentation, and it lead to a mistake earlier. So now it got removed.

glouppe commented:
@arjoly What is your opinion now on this?

arjoly commented:
I think that my remarks still hold. Have I missed something?

betatim commented:
Could you elaborate on that a bit? I think at this point I don't understand what exactly it is you are worried about.

betatim committed:
oob_score will only be calculated if warm_start=False

betatim committed:
Keep estimators_samples_ and estimators_features_ across warmstarts

Keep these attributes as well as estimators_ when adding more
estimators via warm_start=True

arjoly commented:
There is a travis error.

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -65,7 +65,7 @@ def _parallel_build_estimators(n_estimators, ensemble, X, y, sample_weight,
 
         random_state = check_random_state(seeds[i])
-        seed = check_random_state(random_state.randint(MAX_INT))
+        seed = random_state.randint(MAX_INT)
```
Can you revert previous line?

betatim replied:
I changed it to this after discussing with @glouppe

arjoly replied:
Can you summarize the result of the discussion? I though that passing `random_state` to joblib parallel has to be avoided to avoid serializing a random_state object (which is (was?) costly).

betatim replied:
As it is now `seed` is an integer, if we revert the change it will be a `random_state` object again. The `seeds` argument to `_parallel_build_estimators()` is indeed an array of `int`s, this is the argument that is being serialised by joblib parallel no?

Vague discussion summary: I asked at the sprint why my `sklearn.ensemble.tests.test_bagging.test_warm_start` was failing, we discussed a bit and couldn't find a good reason and as part of it decided that it seemed weird to have `seed` be a `random_state` derived from a `random_state` by seeding it, drawing a random int and using that as a seed. My memory is quite vague, so we should wait for @glouppe to say something.

Observation: reverting the change causes `sklearn.ensemble.tests.test_bagging.test_warm_start` to fail. Is L596 of `test_bagging.py` the right way to check that the trees have the same random state?

glouppe replied:
@arjoly The point is actually to avoid passing a random state object.

glouppe replied:
and also to ensure that trees built in a single fit or trees built with multiple small calls with warm_restart will be the same (i.e. generated from the same seed)

arjoly replied:
Sorry I failed to see the bigger picture with the missing lines of the diff. Looks good.

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -218,10 +219,14 @@ def __init__(self,
         self.verbose = verbose
 
+        self.estimators_samples_ = []
+        self.estimators_features_ = []
```
Can you remove those ? Api-wise, it has been decided that fitted attribute should not be set during initialisation.

betatim replied:
I added them to be similar to `self.estimators_` which is set in `BaseEnsemble.__init__`.

Where is a good place to set them instead? Main reason I added them is to make the reset-if-allocated-handling in `fit()` similar for `estimators_`, `estimators_samples_` and `estimators_features_`.

arjoly replied:
Those should be set in the fit. You might want to use `hasattr` to check if those variables are already allocated.

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -278,13 +283,38 @@ def fit(self, X, y, sample_weight=None):
-        self.estimators_ = None
+        if self.warm_start and self.oob_score:
+            raise ValueError("Out of bag estimate only available"
+                             " if warm_start=False")
```
Thanks !

arjoly commented on sklearn/ensemble/bagging.py:
```diff
@@ -298,13 +328,16 @@ def fit(self, X, y, sample_weight=None):
             t[2] for t in all_results))
 
+        if hasattr(self, "oob_score_") and self.warm_start:
+            del self.oob_score_
```
I would do that around line 290-294 to handle fitted attribute at the same place.

arjoly commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -571,6 +572,99 @@ def test_bagging_sample_weight_unsupported_but_passed():
+    clf.fit(X, y)
+    
+    with assert_raises(AttributeError):
+        print clf.oob_score_
```
You can use something like `assert_raise(AttributeError, getattr, clf, "oob_score_")` to avoid printing things.

betatim commented:
Investigating the travis errors, related to adding `estimators_{features,samples}_` in the init I think. Sorry about leaving it in a broken state but I had to stop work mid update :-/

betatim committed:
Better assert_raises use, API compliance, better ordering

Removed the need for printing attributes to raise
exceptions in the tests

Removed unnecessary setting of attributes in the
initialisation.

Moved check for warm_start and oob_score

arjoly commented:
LGTM ! Thanks @betatim for your patience.

betatim commented:
Cool beans! Thanks for your patience and putting up with all the questions and missteps from a contributor n00b.

arjoly commented:
Merge by rebase !! :beers:
